src/NativeFunctionCall.ts

```diff
@@ -0,0 +1,436 @@
+import {StoryException} from './StoryException';
+import {Void} from './Void';
+import {Path} from './Path';
+import {InkList, InkListItem} from './InkList';
+import {InkObject} from './Object';
+import {asOrNull, asOrThrows} from './TypeAssertion';
+import {throwNullException} from './NullException';
+
+type BinaryOp<T> = (left: T, right: T) => any;
+type UnaryOp<T> = (val: T) => any;
+
+export class NativeFunctionCall extends InkObject{
+	// tslint:disable:variable-name
+	public static readonly Add: string 		= '+';
+	public static readonly Subtract: string = '-';
+	public static readonly Divide: string   = '/';
+	public static readonly Multiply: string = '*';
+	public static readonly Mod: string      = '%';
+	public static readonly Negate: string   = '_';
+	public static readonly Equal: string    = '==';
+	public static readonly Greater: string  = '>';
+	public static readonly Less: string     = '<';
+	public static readonly GreaterThanOrEquals: string = '>=';
+	public static readonly LessThanOrEquals: string = '<=';
+	public static readonly NotEquals: string   = '!=';
+	public static readonly Not: string      = '!';
+	public static readonly And: string      = '&&';
+	public static readonly Or: string       = '||';
+	public static readonly Min: string      = 'MIN';
+	public static readonly Max: string      = 'MAX';
+	public static readonly Has: string      = '?';
+	public static readonly Hasnt: string    = '!?';
+	public static readonly Intersect: string = '^';
+	public static readonly ListMin: string   = 'LIST_MIN';
+	public static readonly ListMax: string   = 'LIST_MAX';
+	public static readonly All: string       = 'LIST_ALL';
+	public static readonly Count: string     = 'LIST_COUNT';
+	public static readonly ValueOfList: string = 'LIST_VALUE';
+	public static readonly Invert: string    = 'LIST_INVERT';
+	// tslint:enable:variable-name
+
+	public static CallWithName(functionName: string){
+		return new NativeFunctionCall(functionName);
+	}
+
+	public static CallExistsWithName(functionName: string){
+		this.GenerateNativeFunctionsIfNecessary();
+		return this._nativeFunctions!.get(functionName);
+	}
+
+	get name(){
+		if (this._name === null) return throwNullException('NativeFunctionCall._name');
+		return this._name;
+	}
+	set name(value: string){
+		this._name = value;
+		if( !this._isPrototype ) {
+			if (NativeFunctionCall._nativeFunctions === null) throwNullException('NativeFunctionCall._nativeFunctions');
+			else this._prototype = NativeFunctionCall._nativeFunctions.get(this._name) || null;
+		}
+	}
+	public _name: string | null = null;
+
+	get numberOfParameters(){
+		if (this._prototype) {
+			return this._prototype.numberOfParameters;
+		} else {
+			if (this._numberOfParameters === null) return throwNullException('NativeFunctionCall._numberOfParameters');
+			return this._numberOfParameters;
+		}
+	}
+	set numberOfParameters(value: number){
+		this._numberOfParameters = value;
+	}
+	public _numberOfParameters: number | null = null;
```

Comment:
I'm not sure this should be allowed to be `null`, since in the original code the default value will be 0. Did you run into issues with a non nullable `_numberOfParameters`?

Reply:
Nah I thought the default value was `null`, not `0`. i'll change that!